### PR TITLE
pkg/ipcache: fix deletion of entries from ipcache when event received from kvstore

### PR DESCRIPTION
**Pull request by ianvernon**
The value is only populated for create events; for delete events, it is not
populated. Thus, added a new function which extracts the IP from the key; this
is similar to what is done for `pkg/kvstore/allocator`. Add unit tests for this
conversion as well.

Also do the following:

* Update code-comment for EventTypeListDone to be more accurate.
* Add event type to log message were we cannot unmarshal data from the key-value store.

Signed-off by: Ian Vernon <ian@cilium.io>

Fixes: #3206

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

File: pkg/ipcache/ipcache.go [hunk 1]
```diff
@@ -189,6 +191,21 @@ const (
 	Delete CacheModification = "Delete"
 )
 
+func keyToIP(key string) net.IP {
+	fullPath := path.Join(path.Join(IPIdentitiesPath, AddressSpace))
```
Comment (tgraf) [inline pkg/ipcache/ipcache.go]: What is the nested `path.Join` for?

Comment (tgraf) [inline pkg/ipcache/ipcache.go]: Call this `requiredPrefix`?

File: pkg/ipcache/ipcache.go [hunk 2]
```diff
@@ -189,6 +191,21 @@ const (
...
+	if !strings.HasPrefix(key, fullPath) {
+		log.WithFields(logrus.Fields{"key": key, "prefix": IPIdentitiesPath}).Warning("Found invalid key outside of prefix")
+		return nil
+	}
+
+	suffix := strings.TrimPrefix(key, fullPath)
```
Comment (tgraf) [inline pkg/ipcache/ipcache.go]: Add the `/` to the `requiredPrefix` so gets trimmed away? I don't see a case where we have a valid suffix without a `/`.

File: pkg/ipcache/ipcache.go [hunk 3]
```diff
@@ -189,6 +191,21 @@ const (
 	Delete CacheModification = "Delete"
 )
 
+func keyToIP(key string) net.IP {
```
Comment (tgraf) [inline pkg/ipcache/ipcache.go]: Return an `error` as well so you can give details why the convert failed

File: pkg/ipcache/ipcache.go [hunk 4]
```diff
@@ -242,24 +261,48 @@ func ipIdentityWatcher(owner IPIdentityMappingOwner) {
 					envoy.NetworkPolicyHostsCache.Upsert(envoy.NetworkPolicyHostsTypeURL, ipIDPair.ID.StringID(), &envoyAPI.NetworkPolicyHosts{Policy: uint64(ipIDPair.ID), HostAddresses: ipStrings}, false)
 				}
 			case kvstore.EventTypeDelete:
+				// Synchronize local caching of endpoint IP to ipIDPair mapping with
+				// operation key-value store has informed us about.
+
+				convertedKey := keyToIP(event.Key)
```
Comment (tgraf) [inline pkg/ipcache/ipcache.go]: Retrieve the error here and log it instead of the generic `"unable to convert key to IP"`

**Review by tgraf — CHANGES_REQUESTED** [review]

File: pkg/ipcache/ipcache.go [hunk 5]
```diff
@@ -242,24 +261,48 @@ func ipIdentityWatcher(owner IPIdentityMappingOwner) {
...
+
+				cachedIdentity, exists := IPIdentityCache.LookupByIP(ipStr)
+
 				if exists {
+
+					// Perform lookup first to get list of identities.
+					identityToDelete, exists := IPIdentityCache.LookupByIP(ipStr)
+					ipIDPair.ID = identityToDelete
+
+					endpointIPs, exists := IPIdentityCache.LookupByIdentity(identityToDelete)
```
Comment (raybejjani) [inline pkg/ipcache/ipcache.go]: This exists overwrites the one 3 lines above, I think. Is that intended?

File: pkg/ipcache/ipcache.go (hunk 1 above)
Comment (ianvernon) [reply pkg/ipcache/ipcache.go]: Will be fixed in next upload.

File: pkg/ipcache/ipcache.go (hunk 5 above)
Comment (ianvernon) [reply pkg/ipcache/ipcache.go]: Yes - the first exists isn't checked so I'll make it an underscore to explicitly show that I'm ignoring it.

**Comment by ianvernon** [comment]
test-me-please

File: pkg/ipcache/ipcache.go [hunk 6]
```diff
@@ -27,6 +29,7 @@ import (
 	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 
+	"fmt"
```
Comment (tgraf) [inline pkg/ipcache/ipcache.go]: Move this

File: pkg/ipcache/ipcache.go [hunk 7]
```diff
@@ -189,6 +192,25 @@ const (
...
 )
 
+func keyToIP(key string) (net.IP, error) {
+	requiredPrefix := fmt.Sprintf("%s/", path.Join(IPIdentitiesPath, AddressSpace))
+	if !strings.HasPrefix(key, requiredPrefix) {
+		return nil, fmt.Errorf("Found invalid key %s outside of prefix %s", key, IPIdentitiesPath)
+	}
+
+	suffix := strings.TrimPrefix(key, requiredPrefix)
+	if suffix[0] == '/' {
```
Comment (tgraf) [inline pkg/ipcache/ipcache.go]: Can this ever happen if you trim the prefix with `/`?

File: pkg/ipcache/ipcache.go [hunk 8]
```diff
@@ -189,6 +192,25 @@ const (
...
+		suffix = suffix[1:]
+	}
+
+	parsedIP := net.ParseIP(suffix)
+	if parsedIP == nil {
+		return nil, fmt.Errorf("unable to parse IP from key %s", key)
```
Comment (tgraf) [inline pkg/ipcache/ipcache.go]: Add the suffix here and not the whole key, we want to know what was attempted to be parsed.

**Review by tgraf — CHANGES_REQUESTED** [review]

File: pkg/ipcache/ipcache.go (hunk 7 above)
Comment (ianvernon) [reply pkg/ipcache/ipcache.go]: Good point, will remove.

File: pkg/ipcache/ipcache.go (hunk 8 above)
Comment (ianvernon) [reply pkg/ipcache/ipcache.go]: Fixed.

**Comment by ianvernon** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

File: pkg/ipcache/ipcache.go [hunk 9]
```diff
@@ -198,34 +217,36 @@ func ipIdentityWatcher(owner IPIdentityMappingOwner) {
...
 				cacheChanged      bool
 				cacheModification CacheModification
 				ipIDPair          identity.IPIdentityPair
+				cachedIdentity    identity.NumericIdentity
 			)
 
-			// Key and value are empty for ListDone event types, so just continue.
-			// See GH-3159.
+			// Key and value are empty for ListDone event types; do not try
+			// to unmarshal key or value.
```
Comment (raybejjani) [inline pkg/ipcache/ipcache.go]: Could this be a case in the switch below? I realise it was separate before but the unmarshal between the two is now somewhere else.

File: pkg/ipcache/ipcache.go [hunk 10]
```diff
@@ -242,24 +263,48 @@ func ipIdentityWatcher(owner IPIdentityMappingOwner) {
...
+
+				ipIDPair.IP = convertedKey
+				ipStr := convertedKey.String()
+
+				cachedIdentity, exists := IPIdentityCache.LookupByIP(ipStr)
+
 				if exists {
+
+					// Perform lookup first to get list of identities.
+					identityToDelete, _ := IPIdentityCache.LookupByIP(ipStr)
```
Comment (raybejjani) [inline pkg/ipcache/ipcache.go]: Is this just `cachedIdentity`? I imagine that's why we don't need to check the exists which I asked about earlier.

File: pkg/ipcache/ipcache.go [hunk 11]
```diff
@@ -242,24 +263,48 @@ func ipIdentityWatcher(owner IPIdentityMappingOwner) {
...
+					ipIDPair.ID = identityToDelete
+
+					endpointIPs, exists := IPIdentityCache.LookupByIdentity(identityToDelete)
```
Comment (raybejjani) [inline pkg/ipcache/ipcache.go]: would naming this `exists` `identityInCache` (or something) give line 295 more context?

**Review by raybejjani — APPROVED** [review]

File: pkg/ipcache/ipcache.go (hunk 9 above)
Comment (ianvernon) [reply pkg/ipcache/ipcache.go]: Good point, this is cleaner.

File: pkg/ipcache/ipcache.go (hunk 11 above)
Comment (ianvernon) [reply pkg/ipcache/ipcache.go]: Good idea. Will update both `exists` variable names to be more descriptive; it makes the code much more readable.

**Comment by ianvernon** [comment]
test-me-please

**Review by raybejjani — APPROVED** [review]
Very A+1 thanks for adding the comments and the renames!

**Pull request merged**